### PR TITLE
Avoid compiling user services code if none are configured

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Avoid compiling user services code if none are configured

It looks like we can ifdef some of this out if there are no services configured

I'm not 100% sure this isn't used for something else though

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/api/api_pb2_service.cpp [hunk 1]
```diff
@@ -607,12 +607,14 @@ bool APIServerConnectionBase::read_message(uint32_t msg_size, uint32_t msg_type,
       break;
     }
     case 42: {
+#ifdef USE_API_USER_SERVICES
```
Comment (jesserockz) [inline esphome/components/api/api_pb2_service.cpp:610]: These files are generated and cannot be manually ifdef'd like this...
See the first line of the file =)

File: esphome/components/api/api_pb2_service.h [hunk 2]
```diff
@@ -93,7 +93,9 @@ class APIServerConnectionBase : public ProtoService {
   bool send_get_time_response(const GetTimeResponse &msg);
   virtual void on_get_time_response(const GetTimeResponse &value){};
   bool send_list_entities_services_response(const ListEntitiesServicesResponse &msg);
+#ifdef USE_API_USER_SERVICES
```
Comment (jesserockz) [inline esphome/components/api/api_pb2_service.h:96]: Same here, cant manually change this file

File: esphome/components/api/custom_api_device.h [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 namespace esphome {
 namespace api {
+#ifdef USE_API_USER_SERVICES
```
Comment (jesserockz) [inline esphome/components/api/custom_api_device.h:9]: This is a class that a user can use in a custom component, we have no way of knowing if its used at config validation / code generation time

**Review by jesserockz — CHANGES_REQUESTED** [review]
Because of custom user services via included headers, I dont think this is a safe operation to do...

**Comment by bdraco** [comment]
This one is probably not worth pursuing.  We need more runtime heap memory and this isn’t going to make much difference

**Pull request closed without merging**